internal/service/dataexchange/event_action.go

```diff
@@ -0,0 +1,416 @@
+	if diags.HasError() {
+		resp.Diagnostics.Append(diags...)
+		return
+	}
+
+	out, err := conn.CreateEventAction(ctx, input)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			create.ProblemStandardMessage(names.DataExchange, create.ErrActionCreating, ResNameEventAction, "", err),
+			err.Error(),
+		)
+		return
+	}
+	if out == nil || out.Id == nil {
+		resp.Diagnostics.AddError(
+			create.ProblemStandardMessage(names.DataExchange, create.ErrActionCreating, ResNameEventAction, "", nil),
+			errors.New("empty output").Error(),
+		)
+		return
+	}
+
+	data.ARN = types.StringPointerValue(out.Arn)
+	data.ID = types.StringPointerValue(out.Id)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, data)...)
+}
+
+func (r *resourceEventAction) buildCreateInput(ctx context.Context, data resourceEventActionModel) (*dataexchange.CreateEventActionInput, diag.Diagnostics) {
+	resDiag := diag.Diagnostics{}
+
+	input := dataexchange.CreateEventActionInput{
+		Action: &awstypes.Action{},
+		Event:  &awstypes.Event{},
+	}
+
+	if !data.ActionExportRevisionToS3.IsNull() {
+		actionExportRevisionToS3, diags := data.ActionExportRevisionToS3.ToPtr(ctx)
+		if diags.HasError() {
+			resDiag.Append(diags...)
+			return nil, resDiag
+		}
+		input.Action.ExportRevisionToS3 = &awstypes.AutoExportRevisionToS3RequestDetails{
+			RevisionDestination: &awstypes.AutoExportRevisionDestinationEntry{
+				Bucket:     actionExportRevisionToS3.Bucket.ValueStringPointer(),
+				KeyPattern: actionExportRevisionToS3.KeyPattern.ValueStringPointer(),
+			},
+		}
+
+		if !actionExportRevisionToS3.S3EncryptionType.IsNull() {
+			input.Action.ExportRevisionToS3.Encryption = &awstypes.ExportServerSideEncryption{
+				Type:      actionExportRevisionToS3.S3EncryptionType.ValueEnum(),
+				KmsKeyArn: actionExportRevisionToS3.S3EncryptionKmsKeyArn.ValueStringPointer(),
+			}
+		}
+	}
+
+	if !data.EventRevisionPublished.IsNull() {
+		eventRevisionPublished, diags := data.EventRevisionPublished.ToPtr(ctx)
+		if diags.HasError() {
+			return nil, resDiag
+		}
+		input.Event.RevisionPublished = &awstypes.RevisionPublished{
+			DataSetId: eventRevisionPublished.DataSetId.ValueStringPointer(),
+		}
+	}
+
+	return &input, resDiag
+}
+
+func (r *resourceEventAction) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	conn := r.Meta().DataExchangeClient(ctx)
+
+	var state resourceEventActionModel
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	out, err := findEventActionByID(ctx, conn, state.ID.ValueString())
+	if tfresource.NotFound(err) {
+		resp.State.RemoveResource(ctx)
+		return
+	}
+	if err != nil {
+		resp.Diagnostics.AddError(
+			create.ProblemStandardMessage(names.DataExchange, create.ErrActionSetting, ResNameEventAction, state.ID.String(), err),
+			err.Error(),
+		)
+		return
+	}
+
+	resp.Diagnostics.Append(r.buildModelFromOutput(ctx, out, &state)...)
```

Comment:
Similar to the comment about, this should use the AutoFlex `Flatten` function rather than custom logic to translate between AWS and Terraform data structures.